### DISCUSSION TITLE
Fully serverless Budibase on AWS

### DISCUSSION THREAD

**Discussion by pontusvision** [Budibase/budibase#3899] [Ideas]
I'm working on a PoC to deploy a fully serverless version of budibase on AWS, and one of the painful issues at the moment is the hard dependency on Redis and CouchDB.  It would be great to be able to abstract that away.  This would enable the use of S3 directly, or possibly DynamoDB or simple DB for storage.

For now, I'll start experimenting with a server-side Pouch DB in a lambda with EFS, limiting the lambda concurrency to one lambda.  Performance won't be great, but then again, it'll likely be good enough for many use cases.

**Comment by shogunpurple** [comment]
Hey @pontusvision,

Sounds like a cool project - would be interested to hear how you get on.

We won't be moving away from redis or Couch any time soon - redis is fundamental to our automations section/job queuing, and couchDB is the core of all things data in budibase. Same for abstracting away the persistence - it's probably not something we will get round to in the forseeable future, and would be a huge change with lots of testing required.

Since couchDB is essentially just an HTTP API - it's actually one of the more serverless friendly databases. Compared to SQL databases for example; you don't have to manage connections, or use something like the fairly expensive AWS aurora to get good serverless support. Regarding redis - something like [upstash](https://upstash.com/) is a nice serverless friendly alternative to standard redis, that might be of interest.

Out of interest, I'd love to see how the pouchdb-server performs running on that setup you mention. The more scalable solution instead of the pouchdb-server lambda would be to run couch and redis on an EBS backed EC2 instance, either using docker or just bare metal. While a lambda with provisioned concurrency can be bumped up to handle the memory and CPU requirements, an EC2 instance is much more robust for this kind of thing - especially if you used reserved instances to keep down costs.

Let me know how it goes!

**Comment by rdraju** [comment]
I would be interested in hearing how this goes as well. Replacing CouchDB with DynamoDB and using Elasticache (Redis) would make deployment in AWS a breeze.

**Comment by pontusvision** [comment]
Hi, It's coming along slowly; I've created a (very volatile) project to run the serverless couchdb against dynamo db here: https://github.com/pontus-vision/pouchdb-serverless/tree/ts-conversion
It's still very much a work in progress, so documentation is all over the place; I'm now getting a very obscure error  from budibase server:
{"message":"no_db_file","status":404}
will keep posted on the progress...

**Comment by nnuguid-locii** [comment]
would it make sense to have Budibase use ECS Fargate + S3 + DynamoDB + Elasticache. This gives a Budibase a truly AWS Manage capability and would be super easy to administer.
